aya/src/sys/netlink.rs

```diff
@@ -277,7 +277,7 @@ struct Request {
 struct TcRequest {
     header: nlmsghdr,
     tc_info: tcmsg,
-    attrs: [u8; 64],
+    attrs: [u8; 292], // 292 is the smallest size that can accommodate 256-bytes-long tc names allowed by the kernel and is divisible by 4 (alignment requirement).
```

Comment:
please:
- put the comment above the field, and wrap at 80 columns
- provide a citation for the claim that the kernel allows 256 bytes, preferably by linking to kernel source code. Please use 7 characters for the kernel SHA e.g. https://github.com/torvalds/linux/blob/343e537/kernel/bpf/verifier.c#L9
- provide a better explanation for the value selected; 256 is divisible by 4, so this doesn't make sense to me.

Reply:
I described it based on what was already discussed in the issue.
Valid points, I'll update the comment accordingly.

Reply:
Yes, github issues - particular closed ones - are not an acceptable repository of knowledge.

Reply:
This comment is incorrect.
Netlink messages can contain any number of attributes.
`attrs: [u8; 64]` is being used as a fixed-size buffer (see: `request_attributes()`)  that these attributes get written into.

Side note: This code could do with some love.

To move this forwards I would suggest you bump this to `[u8; 512]`, add a comment explaining that the buffer should be sized to hold at least `CLS_BPF_NAME_LEN = 256` bytes, and finally add an an integration test for a TC program with a long name.

Reply:
@dave-tucker  `request_attributes()` is being called from three places (i.e. `netlink_qdisc_add_clsact`, `netlink_qdisc_attach` and `netlink_find_filter_with_name`) where `req` is of type `TcRequest`.

There is only one place (inside `netlink_find_filter_with_name`) that an attribute of arbitrary size is being added to `attrs` buffer: `options.write_attr_bytes(TCA_BPF_NAME as u16, prog_name.to_bytes_with_nul())?;` which is the tc name.

Since all other attributes have a fixed sizes, the required size would be `all other attributes + size of name` and we know the maximum size of name allowed by the kernel is 256 bytes.
Based on what I observed `33 bytes` is what all other attributes require so 289 is the total required space for the `attrs` buffer. I added 3 more bytes for alignment requirement (which is not really necessary because `request_attributes` does that anyway).

With all that, Can you please explain why `[u8; 512]` ?

Reply:
> There is only one place (inside netlink_find_filter_with_name) that an attribute of arbitrary size is being added to attrs buffer: options.write_attr_bytes(TCA_BPF_NAME as u16, prog_name.to_bytes_with_nul())?; which is the tc name.

> Since all other attributes have a fixed sizes, the required size would be all other attributes + size of name

IIRC The netlink protocol allows for varying sized attributes. We've just been lucky that the ones we had fit into the size of buffer - IMO this was probably luck, not judgement - and you're unlucky to have hit the issue with the TC name being longer than the buffer we allocated.

The issue with sizing to EXACTLY 289 bytes is that this is going to break again the moment we add another attribute here alongside the name.

`512` isn't outrageously large, but at least gives us some headroom in the interim.

The right fix, which is significantly more work, is to not use a fized size buffer int the first place.

Reply:
I get that now, Thank you for the explanation.
I'll send the patch with a test soon.